File: backend/internal/types/session.go
package types

import "time"

// Session represents a saved workspace state
type Session struct {
	ID          string                 `json:"id"`
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	CreatedAt   time.Time              `json:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at"`
	Workspace   Workspace              `json:"workspace"`
	Metadata    map[string]interface{} `json:"metadata"`
}

// Workspace contains the complete state of all apps and UI
type Workspace struct {
	Apps        []AppSnapshot `json:"apps"`
	FocusedID   *string       `json:"focused_app_id,omitempty"`
	FocusedHash *string       `json:"focused_app_hash,omitempty"` // Hash for restoration by properties
	ChatState   *ChatState    `json:"chat_state,omitempty"`
	UIState     *UIState      `json:"ui_state,omitempty"`
}

// AppSnapshot captures complete app state for restoration
type AppSnapshot struct {
	ID             string                 `json:"id"`
	Hash           string                 `json:"hash"` // Deterministic hash for matching
	Title          string                 `json:"title"`
	UISpec         map[string]interface{} `json:"ui_spec"`
	State          State                  `json:"state"`
	ParentID       *string                `json:"parent_id,omitempty"`
	CreatedAt      time.Time              `json:"created_at"`
	Metadata       map[string]interface{} `json:"metadata"`
	Services       []string               `json:"services"`
	ComponentState map[string]interface{} `json:"component_state,omitempty"`
}

// ChatState captures chat history
type ChatState struct {
	Messages []Message `json:"messages"`
	Thoughts []Thought `json:"thoughts"`
}

// Message represents a chat message
type Message struct {
	Type      string `json:"type"`
	Content   string `json:"content"`
	Timestamp int64  `json:"timestamp"`
}

// Thought represents a thought stream entry
type Thought struct {
	Content   string `json:"content"`
	Timestamp int64  `json:"timestamp"`
}

// UIState captures frontend UI state
type UIState struct {
	GenerationThoughts []string `json:"generation_thoughts,omitempty"`
	GenerationPreview  string   `json:"generation_preview,omitempty"`
	IsLoading          bool     `json:"is_loading"`
	Error              *string  `json:"error,omitempty"`
}

// SessionMetadata contains summary information
type SessionMetadata struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	AppCount    int       `json:"app_count"`
}

// ToMetadata extracts metadata from session
func (s *Session) ToMetadata() SessionMetadata {
	return SessionMetadata{
		ID:          s.ID,
		Name:        s.Name,
		Description: s.Description,
		CreatedAt:   s.CreatedAt,
		UpdatedAt:   s.UpdatedAt,
		AppCount:    len(s.Workspace.Apps),
	}
}

// SessionStats contains session manager statistics
type SessionStats struct {
	TotalSessions int        `json:"total_sessions"`
	LastSaved     *time.Time `json:"last_saved,omitempty"`
	LastRestored  *time.Time `json:"last_restored,omitempty"`
}

File: ui/src/types/session.ts
/**
 * Session Persistence Types
 * Type definitions for workspace session management
 */

export interface Session {
  id: string;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  workspace: Workspace;
  metadata: Record<string, any>;
}

export interface Workspace {
  apps: AppSnapshot[];
  focused_app_id?: string;
  chat_state?: ChatState;
  ui_state?: UIState;
}

export interface AppSnapshot {
  id: string;
  title: string;
  blueprint: Record<string, any>;
  state: string;
  parent_id?: string;
  created_at: string;
  metadata: Record<string, any>;
  services: string[];
  component_state?: Record<string, any>;
}

export interface ChatState {
  messages: Message[];
  thoughts: Thought[];
}

export interface Message {
  type: string;
  content: string;
  timestamp: number;
}

export interface Thought {
  content: string;
  timestamp: number;
}

export interface UIState {
  generation_thoughts?: string[];
  generation_preview?: string;
  is_loading: boolean;
  error?: string;
}

export interface SessionMetadata {
  id: string;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  app_count: number;
}

export interface SessionStats {
  total_sessions: number;
  last_saved?: string;
  last_restored?: string;
}

export interface SaveSessionRequest {
  name: string;
  description?: string;
  chat_state?: ChatState;
  ui_state?: UIState;
  app_states?: Record<string, any>; // Component state per app
}

export interface SaveSessionResponse {
  success: boolean;
  session: SessionMetadata;
}

export interface ListSessionsResponse {
  sessions: SessionMetadata[];
  stats: SessionStats;
}

export interface RestoreSessionResponse {
  success: boolean;
  workspace: Workspace;
}
